File: engine/adapters/gmode-20050817/runtime_pack.py
"""Compile translated display buffers; never rewrite original VM bytes or offsets."""
import hashlib
import html
import json
from pathlib import Path
import re
import struct
import sys
from vm import parse_bin

sys.path.append(str(Path(__file__).resolve().parents[1] / 'gmode-v2'))
import pack_codec as shared_codec

TAG = re.compile(r'<[^>]+>')


def normalize_added_quote_colors(rows):
    """Keep added outer dialogue punctuation outside source highlighting."""
    if not rows or not rows[0]['text'].startswith('“') or rows[0]['colors'][0] == 0:
        return
    original_quote = rows[0]['source'].startswith(('「', '『', '“', '‘'))
    if original_quote and not rows[0]['text'].startswith('“‘'):
        return
    rows[0]['colors'][0] = 0
    for row in rows:
        close = row['text'].find('”')
        if close >= 0:
            row['colors'][close] = 0
            break

def fullwidth(text):
    result = ''.join(chr(ord(c)+0xfee0) if '!' <= c <= '~' else '\u3000' if c == ' ' else c for c in text)
    if any(ord(c) < 32 or ord(c) > 0xffff or 0xd800 <= ord(c) <= 0xdfff for c in result):
        raise ValueError('Runtime slots require printable BMP characters')
    return result

def target_rows(source, target):
    if TAG.findall(source) != TAG.findall(target):
        raise ValueError('Translated tags differ from source')
    rows = [dict(text='', colors=[], controls=[])]
    color = None
    last_color = 0
    for token in re.split(r'(<[^>]+>)', target):
        row = rows[-1]
        if not token:
            continue
        if token == '<row/>':
            if color is not None: raise ValueError('Unclosed color at row boundary')
            rows.append(dict(text='', colors=[], controls=[]))
        elif token == '</color>':
            if color is None: raise ValueError('Unexpected color close')
            color = None
        elif re.fullmatch(r'<color=\d+>', token):
            if color is not None: raise ValueError('Nested color')
            color = int(token[7:-1])
            last_color = color
            if not 0 <= color <= 255: raise ValueError('Color outside byte range')
        elif re.fullmatch(r'<ctrl=[0-9A-F]{2}/>', token):
            if not row['controls'] or row['controls'][-1]:
                row['text'] += '\u3000'
                row['colors'].append(last_color)
                row['controls'].append(0)
            row['controls'][-1] = int(token[6:8], 16)
        elif token.startswith('<'):
            raise ValueError('Unknown tag '+token)
        else:
            if color is None: raise ValueError('Text outside color')
            text = fullwidth(html.unescape(token))
            row['text'] += text
            row['colors'].extend([color]*len(text))
            row['controls'].extend([0]*len(text))
    if color is not None: raise ValueError('Unclosed color')
    return rows


def plain_segments(source, target):
    """Align plain spans at the original VM's color and control boundaries.

    Formatting is recovered from the original script at runtime. A control with
    no preceding glyph span still gets an empty slot, so consecutive controls
    and deliberately empty translations retain their original event positions.
    """
    if TAG.findall(source) != TAG.findall(target):
        raise ValueError('Translated tags differ from source')
    original = re.split(r'(<[^>]+>)', source)
    translated = re.split(r'(<[^>]+>)', target)
    rows = [[]]
    active = None
    controlled = False
    for src, dst in zip(original, translated):
        if src == '<row/>':
            if active is not None:
                raise ValueError('Unclosed color at row boundary')
            rows.append([])
            controlled = False
        elif re.fullmatch(r'<color=\d+>', src):
            if active is not None:
                raise ValueError('Nested color')
            active = dict(source='', target='')
        elif src == '</color>':
            if active is None:
                raise ValueError('Unexpected color close')
            rows[-1].append(active)
            active = None
            controlled = False
        elif re.fullmatch(r'<ctrl=[0-9A-F]{2}/>', src):
            if active is not None:
                raise ValueError('Control inside color block')
            if not rows[-1] or controlled:
                rows[-1].append(dict(source='', target=''))
            controlled = True
        elif src.startswith('<'):
            raise ValueError('Unknown tag '+src)
        elif src or dst:
            if active is None:
                raise ValueError('Text outside color')
            active['source'] += html.unescape(src)
            active['target'] += fullwidth(html.unescape(dst))
    if active is not None:
        raise ValueError('Unclosed color')
    return rows

def compile_pack(game):
    game = Path(game)
    sys.path.insert(0, str(game/'scripts'))
    from prepare_translation import row_markup
    document = json.loads((game/'work/dialogue-tagged.json').read_text(encoding='utf-8-sig'))
    groups = {}
    for unit in document['units']:
        groups.setdefault(unit['script'], []).append(unit)
    scripts, hashes = [], {}
    report = dict(units=0, displays=0, strings=0, rows=0, ruby_rows=0, ruby_groups=0,
                  ruby_policy='Original Japanese ruby disabled at runtime; original metadata preserved verbatim.', scripts=[])
    for path in sorted((game/'raw').rglob('*.bin')):
        name = path.parent.name+'/'+path.stem
        raw = path.read_bytes()
        parsed = parse_bin(raw)
        digest = hashlib.sha256(raw[parsed['script_offset']:]).digest()
        commands = {c['offset']: c for c in parsed['commands']}
        record = dict(hash=digest, displays=[], strings=[])
        for unit in groups.pop(name, []):
            c = commands[unit['instruction']]
            if c['opcode'] != unit['opcode'] or c['end'] != unit['end']:
                raise ValueError('Stale command '+unit['id'])
            if not unit['target']: raise ValueError('Missing target '+unit['id'])
            if 'argument' in unit:
                arg = c['args'][unit['argument']]
                visible = c['opcode'] in (5,8,80) and arg['type']=='s' or c['opcode']==17 and unit['argument']==1
                if not visible or html.escape(arg['value'], quote=False) != unit['source']:
                    raise ValueError('Stale/non-display argument '+unit['id'])
                record['strings'].append(dict(offset=arg['offset'], opcode=c['opcode'], source=arg['value'], target=fullwidth(html.unescape(unit['target']))))
            else:
                source = '<row/>'.join(row_markup(r) for r in c['rows'])
                if source != unit['source']: raise ValueError('Stale display '+unit['id'])
                rows = target_rows(source, unit['target'])
                segments = plain_segments(source, unit['target'])
                for row, original, spans in zip(rows, c['rows'], segments):
                    if [x for x in row['controls'] if x] != [x for x in original['controls'] if x]:
                        raise ValueError('Control event sequence changed '+unit['id'])
                    row['source'] = original['text']
                    row['segments'] = spans
                    row['ruby'] = dict(indices=original['ruby_indices'], groups=original['rubies'], cells=original['cells'])
                    report['ruby_rows'] += bool(original['rubies'])
                    report['ruby_groups'] += len(original['rubies'])
                normalize_added_quote_colors(rows)
                record['displays'].append(dict(offset=c['offset'], opcode=c['opcode'], rows=rows))
                report['rows'] += len(rows)
            report['units'] += 1
        report['displays'] += len(record['displays'])
        report['strings'] += len(record['strings'])
        report['scripts'].append(dict(name=name, sha256=digest.hex(), displays=len(record['displays']), strings=len(record['strings'])))
        if digest in hashes:
            if hashes[digest] != record: raise ValueError('Conflicting translations for identical Script SHA256: '+name)
        else:
            hashes[digest] = record
            scripts.append(record)
    if groups: raise ValueError('Missing raw scripts: '+str(list(groups)))
    report['unique_scripts'] = len(scripts)
    report['packed_displays'] = sum(len(s['displays']) for s in scripts)
    report['packed_strings'] = sum(len(s['strings']) for s in scripts)
    report['max_row_slots'] = max(len(r['text']) for s in scripts for d in s['displays'] for r in d['rows'])
    return scripts, report

def row_target_markup(row):
    """Lossless row text using compact color runs and inline control events."""
    if len(row['text']) != len(row['colors']) or len(row['text']) != len(row['controls']):
        raise ValueError('Mismatched row text/color/control lengths')
    result, color = [], None
    for glyph, next_color, control in zip(row['text'], row['colors'], row['controls']):
        if not 0 <= next_color <= 255 or not 0 <= control <= 255:
            raise ValueError('Row color/control outside byte range')
        if color != next_color:
            if color is not None:
                result.append('</color>')
            result.append('<color=%d>' % next_color)
            color = next_color
        result.append(html.escape(glyph, quote=False))
        if control:
            result.append('<ctrl=%02X/>' % control)
    if color is not None:
        result.append('</color>')
    return ''.join(result)


def make_translation_pack(scripts, assembly_hash, scratch_hash, ui, localization=None, literals=None,
                          script_names=None):
    """Export the series schema-1 pack with short names and plain text slots.

    Display slots enumerate spans across all rows of a display. Slot -1 denotes
    a direct string argument. Colors, controls and ruby stay in the game script.
    """
    pack = dict(schema=1, gameAssemblySha256=assembly_hash,
                scratchpadSha256=scratch_hash, scripts=[], ui=list(ui),
                localization=list(localization or []), literals=list(literals or []))
    seen = set()
    def append(script, instruction, slot, opcode, source, target):
        identity = (script, instruction, slot)
        if identity in seen:
            raise ValueError('Duplicate translation slot: '+str(identity))
        seen.add(identity)
        pack['scripts'].append(dict(index=len(pack['scripts']), instruction=instruction,
                                    slot=slot, opcode=opcode, script=script,
                                    source=source, target=target))
    for script in scripts:
        digest = script['hash'].hex()
        if not script_names or digest not in script_names:
            raise ValueError('Missing canonical short script name: '+digest)
        name = script_names[digest]
        for display in script['displays']:
            slot = 0
            for row in display['rows']:
                for span in row['segments']:
                    append(name, display['offset'], slot, display['opcode'],
                           span['source'], span['target'])
                    slot += 1
        for value in script['strings']:
            append(name, value['offset'], -1, value['opcode'], value['source'], value['target'])
    if not pack['scripts'] or not pack['ui']:
        raise ValueError('Translation pack requires scripts and UI entries')
    return pack


def encode_translation_pack(pack):
    """Same KBZH schema-1 byte layout as the shared TranslationPackReader.

    The codec moved to the shared layer (gmode-v2/pack_codec.py) so both version
    adapters write the pack the same way; kept here so existing callers and the
    eighth-game build entry keep their imports.
    """
    return shared_codec.encode_translation_pack(pack)


def main():
    raise SystemExit(
        'Build translations.json / translations.bin through the shared build entry: '
        'python games/08-kibu8/scripts/build_bepinex.py'
    )


if __name__ == '__main__': main()

File: engine/adapters/gmode-20050817/RuntimePack.cs
using System;
using System.IO;
using System.Text;
using System.Security.Cryptography;
using System.Collections.Generic;

namespace Kibukawa8.Runtime
{
    public sealed class RuntimeRow
    {
        public string SourceText, Text, RubyJson;
        public byte[] Colors, Controls;
    }
    public sealed class DisplayTranslation
    {
        public int Offset;
        public byte Opcode;
        public RuntimeRow[] Rows;

        public void NormalizeAddedQuoteColors()
        {
            if(Rows.Length==0 || String.IsNullOrEmpty(Rows[0].Text) || Rows[0].Text[0]!='“' || Rows[0].Colors[0]==0) return;
            string source=Rows[0].SourceText ?? "";
            bool originalQuote=source.StartsWith("「") || source.StartsWith("『") || source.StartsWith("“") || source.StartsWith("‘");
            // A quoted source term becomes an inner single quote when the
            // translation adds a surrounding spoken-dialogue quotation.
            if(originalQuote && !Rows[0].Text.StartsWith("“‘")) return;
            Rows[0].Colors[0]=0;
            foreach(var row in Rows)
            {
                int close=row.Text.IndexOf('”');
                if(close>=0) { row.Colors[close]=0;break; }
            }
        }
    }
    public sealed class StringTranslation
    {
        public int Offset;
        public byte Opcode;
        public string Source, Target;
    }
    public sealed class ScriptTranslation
    {
        public readonly Dictionary<int, DisplayTranslation> Displays = new Dictionary<int, DisplayTranslation>();
        public readonly Dictionary<int, StringTranslation> Strings = new Dictionary<int, StringTranslation>();
    }
    // Shared KBZH schema: translations contain plain text. The immutable original
    // scenario supplies row boundaries, palette indices, and control events.
    public sealed class RuntimePack
    {
        public readonly Dictionary<string, ScriptTranslation> Scripts = new Dictionary<string, ScriptTranslation>(StringComparer.Ordinal);
        public Kibu1ZhCN.TranslationPack Data;
        private Dictionary<string,string> names;
        private readonly HashSet<string> bound = new HashSet<string>(StringComparer.Ordinal);
        private readonly Dictionary<string,Dictionary<int,List<Kibu1ZhCN.ScriptEntry>>> entries = new Dictionary<string,Dictionary<int,List<Kibu1ZhCN.ScriptEntry>>>(StringComparer.Ordinal);
        private static readonly Encoding Sjis = Encoding.GetEncoding(932, EncoderFallback.ExceptionFallback, DecoderFallback.ExceptionFallback);
        private Func<byte[],string> originalDecoder;
        private string Decode(byte[] data, int start, int length)
        {
            if (originalDecoder == null) return Sjis.GetString(data,start,length);
            var slice=new byte[length];Array.Copy(data,start,slice,0,length);
            return originalDecoder(slice);
        }
        public static RuntimePack Load(string path, Dictionary<string,string> scriptNames, Func<byte[],string> originalDecoder = null)
        {
            var pack = new RuntimePack { Data = Kibu1ZhCN.TranslationPackReader.Load(path), names = new Dictionary<string,string>(scriptNames,StringComparer.Ordinal), originalDecoder = originalDecoder };
            var known = new HashSet<string>(scriptNames.Values,StringComparer.Ordinal);
            foreach (var entry in pack.Data.scripts)
            {
                if (entry.script == null || !known.Contains(entry.script) || entry.instruction < 0 || entry.opcode < 0 || entry.opcode > 255 || entry.target == null || entry.source == null)
                    throw new InvalidDataException("Invalid VM translation entry");
                foreach(char c in entry.target)
                    if(c<32 || Char.IsSurrogate(c) || c=='<' || c=='>') throw new InvalidDataException("Translation must contain plain printable text");
                ScriptTranslation script;
                if (!pack.Scripts.TryGetValue(entry.script, out script))
                {
                    pack.Scripts.Add(entry.script, script = new ScriptTranslation());
                    pack.entries.Add(entry.script,new Dictionary<int,List<Kibu1ZhCN.ScriptEntry>>());
                }
                if (entry.slot == -1)
                {
                    if (entry.opcode!=5 && entry.opcode!=8 && entry.opcode!=17 && entry.opcode!=80) throw new InvalidDataException("Invalid string opcode");
                    script.Strings.Add(entry.instruction, new StringTranslation { Offset=entry.instruction, Opcode=(byte)entry.opcode, Source=entry.source, Target=entry.target });
                    continue;
                }
                if (entry.slot < 0 || (entry.opcode!=255 && entry.opcode!=72 && entry.opcode!=75 && entry.opcode!=120)) throw new InvalidDataException("Invalid display slot/opcode");
                List<Kibu1ZhCN.ScriptEntry> display;
                if(!pack.entries[entry.script].TryGetValue(entry.instruction,out display))
                    pack.entries[entry.script].Add(entry.instruction,display=new List<Kibu1ZhCN.ScriptEntry>());
                if(entry.slot!=display.Count || (display.Count!=0 && entry.opcode!=display[0].opcode)) throw new InvalidDataException("Unordered or duplicate display span");
                display.Add(entry);
            }
            return pack;
        }
        public ScriptTranslation Bind(byte[] original)
        {
            string hash;
            using(SHA256 sha=SHA256.Create()) hash=BitConverter.ToString(sha.ComputeHash(original)).Replace("-","").ToLowerInvariant();
            string name; ScriptTranslation script;
            if(!names.TryGetValue(hash,out name) || !Scripts.TryGetValue(name,out script)) return null;
            if(bound.Contains(name)) return script;
            // Stage all reconstructed displays before publishing a script binding.
            var displays=new Dictionary<int,DisplayTranslation>();
            foreach(var pair in entries[name])
            {
                int pos=pair.Key;
                byte op=Take(original,ref pos);
                if(op!=pair.Value[0].opcode) throw new InvalidDataException("Original display opcode mismatch");
                int count=1;
                if(op==255) { count=Take(original,ref pos);Take(original,ref pos); }
                var rows=new RuntimeRow[count]; int slot=0; byte lastColor=0;
                for(int i=0;i<count;i++) rows[i]=ReadRow(original,ref pos,op,pair.Value,ref slot,ref lastColor);
                if(slot!=pair.Value.Count) throw new InvalidDataException("Unexpected translated display spans");
                var display=new DisplayTranslation {Offset=pair.Key,Opcode=op,Rows=rows};
                display.NormalizeAddedQuoteColors();
                displays.Add(pair.Key,display);
            }
            foreach(var item in script.Strings.Values)
            {
                int end=item.Offset;
                while(end<original.Length && original[end]!=0) end++;
                if(end==original.Length || Decode(original,item.Offset,end-item.Offset)!=item.Source)
                    throw new InvalidDataException("Original string source mismatch");
            }
            foreach(var pair in displays) script.Displays.Add(pair.Key,pair.Value);
            bound.Add(name);
            return script;
        }
        private static byte Take(byte[] data,ref int pos)
        {
            if(pos<0 || pos>=data.Length) throw new InvalidDataException("Truncated original display");
            return data[pos++];
        }
        private static void Skip(byte[] data,ref int pos,int count)
        {
            if(count<0 || pos<0 || count>data.Length-pos) throw new InvalidDataException("Truncated original display plane");
            pos+=count;
        }
        private sealed class Span { public string Source; public byte Color,Control; }
        private RuntimeRow ReadRow(byte[] data,ref int pos,byte op,List<Kibu1ZhCN.ScriptEntry> entries,ref int slot,ref byte lastColor)
        {
            int n=Take(data,ref pos), textOffset=pos; Skip(data,ref pos,n*2);
            string source=Decode(data,textOffset,n*2);
            if(op==255 || op==75) Skip(data,ref pos,n);
            int colorsOffset=pos; Skip(data,ref pos,n);
            int controlsOffset=pos; bool hasControls=op==255 || op==120;
            if(hasControls) Skip(data,ref pos,n);
            if(op==255 || op==75)
            {
                int rubies=Take(data,ref pos);
                for(int i=0;i<rubies;i++) { Take(data,ref pos); int cells=Take(data,ref pos); Skip(data,ref pos,cells*2); }
            }
            var spans=new List<Span>(); var pending=new StringBuilder(); byte color=0;
            for(int i=0;i<n;i++)
            {
                byte next=data[colorsOffset+i];
                if(next!=color && pending.Length!=0) { spans.Add(new Span {Source=pending.ToString(),Color=color});pending.Length=0; }
                color=next;
                pending.Append(Decode(data,textOffset+i*2,2).Replace("\uf8f3",""));
                byte control=hasControls?data[controlsOffset+i]:(byte)0;
                if(control!=0)
                {
                    if(pending.Length!=0) {spans.Add(new Span {Source=pending.ToString(),Color=color});pending.Length=0;}
                    if(spans.Count==0 || spans[spans.Count-1].Control!=0) spans.Add(new Span {Source="",Color=color});
                    spans[spans.Count-1].Control=control;
                }
            }
            if(pending.Length!=0) spans.Add(new Span {Source=pending.ToString(),Color=color});
            var text=new StringBuilder();var colors=new List<byte>();var controls=new List<byte>();
            foreach(var span in spans)
            {
                if(slot>=entries.Count || entries[slot].source!=span.Source) throw new InvalidDataException("Original display span mismatch at slot "+slot+" expected UTF16="+(slot<entries.Count?BitConverter.ToString(Encoding.Unicode.GetBytes(entries[slot].source)):"missing")+" actual UTF16="+BitConverter.ToString(Encoding.Unicode.GetBytes(span.Source)));
                string target=entries[slot++].target;
                if(span.Source.Length!=0) lastColor=span.Color;
                text.Append(target);
                for(int i=0;i<target.Length;i++) {colors.Add(lastColor);controls.Add(0);}
                if(span.Control!=0)
                {
                    if(controls.Count==0 || controls[controls.Count-1]!=0) {text.Append('\u3000');colors.Add(lastColor);controls.Add(0);}
                    controls[controls.Count-1]=span.Control;
                }
            }
            return new RuntimeRow {SourceText=source,Text=text.ToString(),Colors=colors.ToArray(),Controls=controls.ToArray(),RubyJson="{}"};
        }
    }
}
